### PR TITLE
[SPARK-37523][SQL] Support optimize skewed partitions in Distribution and Ordering if numPartitions is not specified

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
Support optimize skewed partitions in Distribution and Ordering if numPartitions is not specified

### Why are the changes needed?
When doing repartition in distribution and sort, we will use Rebalance operator instead of RepartitionByExpression to optimize skewed partitions when
1. numPartitions is not specified by the data source, and
2. sortOrder is specified. This is because the requested distribution needs to be guaranteed, which can only be achieved by using RangePartitioning, not HashPartitioning.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing and new tests

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 1]
```diff
@@ -1440,6 +1441,14 @@ case class RepartitionByExpression(
 }
 
 object RepartitionByExpression {
+  def apply(
+      partitionExpressions: Seq[Expression],
+      child: LogicalPlan,
+      numPartitions: Int,
+      adapativePartition: Boolean): RepartitionByExpression = {
```
Comment (sunchao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:1448]: `adapative` -> `adaptive`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala [hunk 2]
```diff
@@ -37,15 +37,15 @@ object DistributionAndOrderingUtils {
...
+          RepartitionByExpression(distribution, query, numPartitions)
         } else {
-          conf.numShufflePartitions
+          RepartitionByExpression(distribution, query, conf.numShufflePartitions, true)
         }
         // the conversion to catalyst expressions above produces SortOrder expressions
         // for OrderedDistribution and generic expressions for ClusteredDistribution
         // this allows RepartitionByExpression to pick either range or hash partitioning
-        RepartitionByExpression(distribution, query, finalNumPartitions)
+
```
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:48]: should we move the comments up?

**Review by sunchao — COMMENTED** [review]
This looks reasonable to me.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 1 above)
Comment (huaxingao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Fixed

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala (hunk 2 above)
Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala]: Fixed. Thanks

**Comment by HyukjinKwon** [comment]
cc @cloud-fan and @maryannxue FYI

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 3]
```diff
@@ -1405,7 +1405,8 @@ case class Repartition(numPartitions: Int, shuffle: Boolean, child: LogicalPlan)
 case class RepartitionByExpression(
     partitionExpressions: Seq[Expression],
     child: LogicalPlan,
-    optNumPartitions: Option[Int]) extends RepartitionOperation {
+    optNumPartitions: Option[Int],
+    adaptivePartition: Boolean = false) extends RepartitionOperation {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:1409]: maybe `userSpecified: Boolean`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 4]
```diff
@@ -819,6 +819,8 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
       case r: logical.RepartitionByExpression =>
         val shuffleOrigin = if (r.partitionExpressions.isEmpty && r.optNumPartitions.isEmpty) {
           REBALANCE_PARTITIONS_BY_NONE
+        } else if (!r.userSpecified) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:822]: I think we can make the logic clearer here:
```
if (numPart is empty) {
  if (partExpr is empty) {
    REBALANCE_PARTITIONS_BY_NONE
  } else if (userSpecified) {
    REPARTITION_BY_COL
  } else {
    REBALANCE_PARTITIONS_BY_COL
  }
} else {
  REPARTITION_BY_NUM
}
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala [hunk 5]
```diff
@@ -37,15 +37,16 @@ object DistributionAndOrderingUtils {
       }
 
       val queryWithDistribution = if (distribution.nonEmpty) {
-        val finalNumPartitions = if (numPartitions > 0) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:40]: We can make the change smaller, just pass `userSpecified = false`

File: sql/core/src/test/scala/org/apache/spark/sql/connector/WriteDistributionAndOrderingSuite.scala [hunk 6]
```diff
@@ -63,36 +65,51 @@ class WriteDistributionAndOrderingSuite
   private val resolver = conf.resolver
 
   test("ordered distribution and sort with same exprs: append") {
-    checkOrderedDistributionAndSortWithSameExprs("append")
+    Seq(false, true).foreach { skew =>
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/WriteDistributionAndOrderingSuite.scala:68]: If every test needs to run with both skew and non-skew, I'd suggest we keep the old test suite unchanged, and add a new `WriteDistributionAndOrderingWithSkewSuite` which extends the old suite.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/WriteDistributionAndOrderingSuite.scala [hunk 7]
```diff
@@ -685,54 +788,88 @@ class WriteDistributionAndOrderingSuite
...
+          val resolvedOrdering = p.ordering.map(resolveAttrs(_, plan))
+          p.copy(ordering = resolvedOrdering.asInstanceOf[Seq[catalyst.expressions.SortOrder]])
+        case p: physical.HashPartitioning =>
+          val resolvedExprs = p.expressions.map(resolveAttrs(_, plan))
+          p.copy(expressions = resolvedExprs)
+        case other => other
+      }
+      assert(actualPartitioning == expectedPartitioning, "partitioning must match")
+    } else {
+      assert(actualPartitioning.numPartitions > conf.numShufflePartitions)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/WriteDistributionAndOrderingSuite.scala:871]: can we collect the `AQEShuffleReadExec` in the plan and make sure we do split the skewed partitions?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 8]
```diff
@@ -1405,7 +1405,8 @@ case class Repartition(numPartitions: Int, shuffle: Boolean, child: LogicalPlan)
 case class RepartitionByExpression(
     partitionExpressions: Seq[Expression],
     child: LogicalPlan,
-    optNumPartitions: Option[Int]) extends RepartitionOperation {
+    optNumPartitions: Option[Int],
+    userSpecified: Boolean = true) extends RepartitionOperation {
```
Comment (ulysses-you) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:1409]: I'm afraid some rules will remove `RepartitionByExpression` without check the `userSpecified`, e.g. `RepartitionOperation` which may cause the output partitioning unexpected. That's why we add a new operator `RebalancePartitions` before. If the only difference is `optNumPartitions`, can't we just reuse `RebalancePartitions` ?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:1409]: good point. Maybe we should update `RebalancePartitions` to support `RangePartitioning`.

Comment (maryannxue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:1409]: A little confused here, doesn't the `Option` part of `optNumPartitions: Option[Int]` already contain the information of `userSpecified`?

**Comment by maryannxue** [comment]
One question based on the PR description:
> Spark should optimize repartition and split the skewed partitions if necessary.

If we split a partition, the output would not conform to the specified partitioning any more, right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 8 above)
Comment (maryannxue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:1409]: +1 on the idea of using `RebalancePartitions` instead. We should not mess up the semantics of `RepartitionByExpression`.

Comment (huaxingao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:1409]: I quickly tried `RebalancePartitions`, but it introduced one extra sort and shuffle. Is this OK?

Comment (huaxingao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:1409]: cc @aokolnychyi @RussellSpitzer

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 9]
```diff
@@ -819,6 +819,8 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
...
+          REBALANCE_PARTITIONS_BY_COL
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:823]: `REBALANCE_PARTITIONS_BY_COL` seems more suitable to be from `RebalancePartitions`? Although there is `REBALANCE_PARTITIONS_BY_NONE` here too.

After rebalancing, do we still guarantee partitioning requirement by the partition expressions? Seems not to me as `RebalancePartitions` said:

> ... It also try its best to partition the child output by `partitionExpressions`....

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 10]
```diff
@@ -1466,7 +1475,20 @@ case class RebalancePartitions(
   def partitioning: Partitioning = if (partitionExpressions.isEmpty) {
     RoundRobinPartitioning(conf.numShufflePartitions)
   } else {
-    HashPartitioning(partitionExpressions, conf.numShufflePartitions)
+    val (sortOrder, nonSortOrder) = partitionExpressions.partition(_.isInstanceOf[SortOrder])
+    require(sortOrder.isEmpty || nonSortOrder.isEmpty,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:1479]: We can probably have a common trait for this and `RepartitionByExpression`, to share code.

**Comment by cloud-fan** [comment]
Can you update the PR description and fix the code conflicts? Seems we don't need to change `RepartitionByExpression` anymore.

IIUC, the new proposal of this PR is: For distribution required by the v2 sink, we use Rebalance operator instead of `RepartitionByExpression`. This looks reasonable, as it's for performance and best-effort partitioning is good enough. AFAFIK @wangyum is working on something similar for file sources.

We should update the DS v2 API doc though. To mention that the required distribution is just a best effort.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 11]
```diff
@@ -1395,6 +1395,33 @@ case class Repartition(numPartitions: Int, shuffle: Boolean, child: LogicalPlan)
     copy(child = newChild)
 }
 
+trait hasPartitionExpressions extends SQLConfHelper {
+
+  def partitionExpressions: Seq[Expression]
+
+  def optNumPartitions: Option[Int]
+
+  def partitioning: Partitioning = if (partitionExpressions.isEmpty) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:1404]: nit: `protected def getPartitioning` to avoid name conflicts.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:1404]: or simply make this a `val`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 12]
```diff
@@ -1459,15 +1471,10 @@ object RepartitionByExpression {
...
+    child: LogicalPlan) extends UnaryNode with hasPartitionExpressions {
   override def maxRows: Option[Long] = child.maxRows
   override def output: Seq[Attribute] = child.output
-
-  def partitioning: Partitioning = if (partitionExpressions.isEmpty) {
-    RoundRobinPartitioning(conf.numShufflePartitions)
-  } else {
-    HashPartitioning(partitionExpressions, conf.numShufflePartitions)
-  }
+  override def optNumPartitions: Option[Int] = Some(conf.numShufflePartitions)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:1477]: should this be None?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala [hunk 13]
```diff
@@ -36,16 +36,27 @@ object DistributionAndOrderingUtils {
         case _: UnspecifiedDistribution => Array.empty[Expression]
       }
 
+      val (sortOrder, _) = distribution.partition(_.isInstanceOf[SortOrder])
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:39]:
```suggestion
      val sortOrder = distribution.filter(_.isInstanceOf[SortOrder])
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala [hunk 14]
```diff
@@ -36,16 +36,27 @@ object DistributionAndOrderingUtils {
...
+
       val queryWithDistribution = if (distribution.nonEmpty) {
-        val finalNumPartitions = if (numPartitions > 0) {
-          numPartitions
+        // Spark can optimize the partition when
+        // 1. numPartitions is not specified by the data source, and
+        // 2. sortOrder is specified. This is because the requested distribution needs to be
+        // guaranteed, which can only be achieved by using RangePartitioning, not HashPartitioning.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:45]: This looks too strict. Do we have a reason to must keep the hash partitioning? For file sources, we want to shuffle by partition columns to reduce the number of files, but it's just a best effort as a rough hash partitioning also works.

Do we have other special requirements for DS v2?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:45]: Anyway, we need a DS v2 API to let the sink require "rough" hash partitioning. If we don't want to change the existing API, we need to add a new one for it.

**Comment by aokolnychyi** [comment]
Thanks for the PR, @huaxingao. I think it is a great feature and it would be awesome to get it done.

I spent some time thinking about this and have a few questions/proposals.

If I understand correctly, we currently hard-code the number of shuffle partitions in `RepartitionByExpression`, which prohibits both coalescing and skew split optimizations.

It seems reasonable to support cases when the requested distribution is best-effort but I also think there are valid cases when the distribution is required for correctness and it is actually the current API contract. What about extending `RequiredDistributionAndOrdering` to indicate the distribution is not strictly required? We can add some boolean method and default it to keep the existing behavior. If the distribution is required, we can still benefit from coalescing as I think `CoalesceShufflePartitions` and `AQEShuffleReadExec` would keep the original distribution in coalesce cases. That’s already a huge win. We can avoid too small files while keeping the requested distribution.

I also agree about using `RebalancePartitions` when the distribution is not strictly required. What about extending `RebalancePartitions` to also support range partitioning? It currently supports only hash and round-robin. If we make that change, we will be able to remove unnecessary shuffles in the optimizer and keep the original distribution as long as there is no skew and we only coalesce. If there is a skew, an extra shuffle and changed distribution seems like a reasonable overhead.

What does everybody else think?

**Comment by huaxingao** [comment]
Thanks @aokolnychyi for the proposal. I agree that we should support both strictly required distribution and best effort distribution. For best effort distribution, if user doesn't request a specific number of partitions, we will split skewed partitions and coalesce small partitions. For strictly required distribution, if user doesn't request a specific number of partitions, we will coalesce small partitions but we will NOT split skewed partitions since this changes the required distribution.

In interface `RequiresDistributionAndOrdering`, I will add
```
default boolean distributionStrictlyRequired() { return true; }
```
Then in `DistributionAndOrderingUtils`.`prepareQuery`, I will change the code to something like this:
```      
      val queryWithDistribution = if (distribution.nonEmpty) {
        if (!write.distributionStrictlyRequired() && numPartitions == 0) {
          RebalancePartitions(distribution, query)
        } else {
          if (numPartitions > 0) {
            RepartitionByExpression(distribution, query, numPartitions)
          } else {
            RepartitionByExpression(distribution, query, None)
          }
        }
        ...
``` 
Basically, in the best effort case, if the requested numPartitions is 0, we will use `RebalancePartitions` so both `OptimizeSkewInRebalancePartitions` and `CoalesceShufflePartitions` will be applied. In the strictly required distribution case,  if the requested numPartitions is 0, we will use `RepartitionByExpression(distribution, query, None)` so `CoalesceShufflePartitions` will be applied. 

Does this sound correct for every one?

**Comment by huaxingao** [comment]
cc @cloud-fan

**Comment by ulysses-you** [comment]
Looks correct to me. BTW, after Spark3.3 the RebalancePartitions supports specify the initialNumPartition, so the demo code can be:
```scala
val optNumPartitions = if (numPartitions == 0) None else Some(numPartitions)
if (!write.distributionStrictlyRequired()) {
  RebalancePartitions(distribution, query, optNumPartitions)
} else {
  RepartitionByExpression(distribution, query, optNumPartitions)
}
```

**Comment by huaxingao** [comment]
It's a bit too hard to rebase. I will close this PR and the new one is here https://github.com/apache/spark/pull/36644. Thank you all very much for reviewing!
